src/PowerPing/Ping.cs

```diff
@@ -410,24 +411,38 @@ private PingResults SendICMP(PingAttributes attrs)
                         if (rnd.Next(20) == 1) { throw new SocketException(); }
                     }
 
-                    // Wait for response
                     byte[] buffer = new byte[attrs.RecieveBufferSize]; // Ipv4Header.length + IcmpHeader.length + attrs.recievebuffersize
-                    bytesRead = sock.ReceiveFrom(buffer, ref ep);
-                    responseTimer.Stop();
+                    ICMP response;
+                    long responseTimestamp;
+                    do {
+                        // Wait for response
+                        bytesRead = sock.ReceiveFrom(buffer, ref ep);
+                        responseTimestamp = Stopwatch.GetTimestamp();
+
+                        // Store reply packet
+                        response = new ICMP(buffer, bytesRead);
+
+                        // Make sure we received the expected packet, if not, discard it
+                        ushort responseSessionId = BitConverter.ToUInt16(response.message, 0);
+                        ushort responseSequenceNum = BitConverter.ToUInt16(response.message, 2);
+                        if (responseSessionId != sessionId || responseSequenceNum != sequenceNum) {
```

Comment:
Should this validation only happen if response.type == 0 to allow errors (e.g. destination unreachable) to pass through?